pkg/vault/contracts/Vault.sol

```diff
@@ -1296,34 +1305,32 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
                 if (bufferWrappedSurplus > 0) {
                     bufferUnderlyingSurplus = wrappedToken.convertToAssets(bufferWrappedSurplus);
                 }
-                wrappedToken.withdraw(amountOutUnderlying + bufferUnderlyingSurplus, address(this), address(this));
+                vaultUnderlyingDeltaHint = amountOutUnderlying + bufferUnderlyingSurplus;
+                vaultWrappedDeltaHint = wrappedToken.withdraw(vaultUnderlyingDeltaHint, address(this), address(this));
             }
 
             // ERC4626 output should not be trusted, so it's a good practice to measure the amount of
             // deposited and returned tokens.
-            (uint256 vaultUnderlyingDelta, uint256 vaultWrappedDelta) = _updateReservesAfterWrapping(
-                underlyingToken,
-                IERC20(wrappedToken)
-            );
+            _settleUnwrap(underlyingToken, IERC20(wrappedToken), vaultUnderlyingDeltaHint, vaultWrappedDeltaHint);
 
             // Only updates buffer balances if buffer has a surplus of wrapped tokens.
             if (bufferWrappedSurplus > 0) {
                 if (kind == SwapKind.EXACT_IN) {
                     // amountInWrapped is the amountGiven and should not be changed. Any rounding issue that occurs
                     // in the vaultWrappedDelta should be absorbed by the buffer.
-                    bufferWrappedSurplus = vaultWrappedDelta - amountInWrapped;
+                    bufferWrappedSurplus = vaultWrappedDeltaHint - amountInWrapped;
                     // Since bufferWrappedSurplus was unwrapped, the final amountOut needs to discount the underlying
                     // amount that will stay in the buffer. Refresh `bufferUnderlyingSurplus` after external calls
                     // on the wrapped token.
                     bufferUnderlyingSurplus = wrappedToken.convertToAssets(bufferWrappedSurplus);
-                    amountOutUnderlying = vaultUnderlyingDelta - bufferUnderlyingSurplus;
+                    amountOutUnderlying = vaultUnderlyingDeltaHint - bufferUnderlyingSurplus;
                 } else {
                     // If buffer has a wrapped surplus, it unwraps surplus + amountIn, so the final amountIn needs to
                     // be discounted for that.
-                    amountInWrapped = vaultWrappedDelta - bufferWrappedSurplus;
+                    amountInWrapped = vaultWrappedDeltaHint - bufferWrappedSurplus;
                     // amountOutUnderlying is the amountGiven and should not be changed. Any rounding issue that occurs
                     // in the vaultUnderlyingDelta should be absorbed by the buffer.
-                    bufferUnderlyingSurplus = vaultUnderlyingDelta - amountOutUnderlying;
+                    bufferUnderlyingSurplus = vaultUnderlyingDeltaHint - amountOutUnderlying;
                 }
 
                 // In an unwrap operation, the underlying balance of the buffer will increase and the wrapped balance
```

Comment:
Partially outside the diff. Comment above:

// In an unwrap operation, the underlying balance of the buffer will increase and the wrapped balance
// will decrease. To increase the underlying balance, we get the delta amount that was withdrawn
// (vaultUnderlyingDelta) and discount the amount needed for the unwrapping operation
// (amountOutUnderlying). The same logic applies to wrapped balances.